ledger-test-tool/infrastructure/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/SingleParticipantTestContext.scala

```diff
@@ -542,91 +556,113 @@ final class SingleParticipantTestContext private[participant] (
     )
       .map(_.getTransaction)
       .map(tx =>
         tx.transactionId -> tx.events.collect { case Event(Created(e)) =>
-          Primitive.ContractId(e.contractId)
+          companion.toContractId(new ContractId(e.contractId))
         }.head
       )
 
   override def exercise[T](
       party: Party,
-      exercise: Primitive.Update[T],
+      exercise: Update[T],
   ): Future[TransactionTree] =
     submitAndWaitForTransactionTree(
-      submitAndWaitRequest(party, exercise.command)
+      submitAndWaitRequest(party, exercise.commands)
     ).map(_.getTransaction)
 
   override def exercise[T](
       actAs: List[Party],
       readAs: List[Party],
-      exercise: Primitive.Update[T],
+      exercise: Update[T],
   ): Future[TransactionTree] =
     submitAndWaitForTransactionTree(
-      submitAndWaitRequest(actAs, readAs, exercise.command)
+      submitAndWaitRequest(actAs, readAs, exercise.commands)
     ).map(_.getTransaction)
 
   override def exerciseForFlatTransaction[T](
       party: Party,
-      exercise: Primitive.Update[T],
+      exercise: Update[T],
   ): Future[Transaction] =
     submitAndWaitForTransaction(
-      submitAndWaitRequest(party, exercise.command)
+      submitAndWaitRequest(party, exercise.commands)
     ).map(_.getTransaction)
 
-  override def exerciseAndGetContract[T](
+  override def exerciseAndGetContract[TCid <: ContractId[T], T](
       party: Party,
-      exercise: Primitive.Update[Any],
-  ): Future[Primitive.ContractId[T]] =
+      exercise: Update[Exercised[TCid]],
+  )(implicit companion: ContractCompanion[?, TCid, T]): Future[TCid] =
     submitAndWaitForTransaction(
-      submitAndWaitRequest(party, exercise.command)
+      submitAndWaitRequest(party, exercise.commands)
     )
       .map(_.getTransaction)
-      .map(extractContracts)
-      .map(_.head.asInstanceOf[Primitive.ContractId[T]])
+      .map(t => extractContracts(t)(companion))
+      .map(_.head)
 
-  override def exerciseByKey[T](
+  override def exerciseAndGetContractNoDisclose[TCid <: ContractId[?]](
       party: Party,
-      template: Primitive.TemplateId[T],
+      exercise: Update[Exercised[UnitData]],
+  )(implicit companion: ContractCompanion[?, TCid, ?]): Future[TCid] =
+    submitAndWaitForTransaction(
+      submitAndWaitRequest(party, exercise.commands)
+    )
+      .map(_.getTransaction)
+      .map(t => extractContracts(t)(companion))
+      .map(_.head)
+
+  override def exerciseByKey(
+      party: Party,
+      template: Identifier,
       key: Value,
       choice: String,
       argument: Value,
   ): Future[TransactionTree] =
     submitAndWaitForTransactionTree(
       submitAndWaitRequest(
         party,
-        Command.of(
-          Command.Command.ExerciseByKey(
-            ExerciseByKeyCommand(
-              Some(template.unwrap),
-              Option(key),
-              choice,
-              Option(argument),
+        JList.of(
```

Comment:
Why can't we just call `new ExerciseByKeyCommand`?

Reply:
a6c1579f5761a2684995c08ff29be1ef46cc9cf6

Reply:
nice